src/engine/strat_engine/range_alloc.rs

```diff
@@ -0,0 +1,230 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+use std::cmp::min;
+
+use devicemapper::Sectors;
+use ranged_set::RangedSet;
+
+#[derive(Debug)]
+pub struct RangeAllocator {
+    max: u64,
+    used: RangedSet<u64>,
+}
+
+impl RangeAllocator {
+    pub fn new(max: Sectors) -> RangeAllocator {
+        RangeAllocator {
+            max: *max,
+            used: RangedSet::new(),
+        }
+    }
+
+    pub fn new_with_used(max: Sectors, initial_used: &[(Sectors, Sectors)]) -> RangeAllocator {
+        let mut allocator = RangeAllocator {
+            max: *max,
+            used: RangedSet::new(),
+        };
+
+        for &(off, len) in initial_used {
+            allocator.insert_range(off, len)
+        }
+
+        allocator
+    }
+
+    fn check_for_overflow(&self, off: u64, len: u64) {
+        if let None = off.checked_add(len) {
+            panic!("off+len overflows u64");
+        }
+        if off + len > self.max {
+            panic!("off+len greater than max allocator range");
+        }
+    }
+
+    pub fn insert_range(&mut self, off: Sectors, len: Sectors) -> () {
+        if len == Sectors(0) {
+            panic!("insert_range called with length 0");
+        }
+        self.check_for_overflow(*off, *len);
+
+        for val in *off..*off + *len {
+            let inserted = self.used.insert(val);
+            if inserted == false {
+                panic!(format!("inserted value {} already present in RangedSet", val));
+            }
+        }
+    }
+
+    pub fn remove_ranges(&mut self, to_free: &[(Sectors, Sectors)]) -> () {
+        for &(off, len) in to_free {
+            self.check_for_overflow(*off, *len);
+            for val in *off..*off + *len {
+                let removed = self.used.remove(&val);
+                if removed == false {
+                    panic!(format!("tried to remove value {} not present in RangedSet", val));
+                }
+            }
+        }
+    }
+
+    pub fn available(&self) -> Sectors {
+        Sectors((0..self.max).filter(|val| !self.used.contains(val)).count() as u64)
+    }
+
+    pub fn used(&self) -> Sectors {
+        Sectors((0..self.max).filter(|val| self.used.contains(val)).count() as u64)
+    }
+
+    /// Get a list of (offset, length) segments that are in use
+    fn used_ranges(&self) -> Vec<(u64, u64)> {
```

Comment:
You didn't use the devicemapper::Segment type for ranges because it has an un-needed path?  Maybe create a Range type?